src/ovro_lwa_portal/io.py

```diff
@@ -0,0 +1,373 @@
+"""Data loading utilities for OVRO-LWA datasets.
+
+This module provides a unified interface for loading OVRO-LWA data from
+multiple sources including local paths, remote URLs, and DOI identifiers.
+"""
+
+from __future__ import annotations
+
+import logging
+import re
+import warnings
+from pathlib import Path
+from typing import Any
+from urllib.parse import urlparse
+
+import xarray as xr
+
+__all__ = ["open_dataset", "DataSourceError"]
+
+logger = logging.getLogger(__name__)
+
+# DOI pattern: matches both "doi:10.xxxx/xxxxx" and "10.xxxx/xxxxx"
+DOI_PATTERN = re.compile(r"^(?:doi:)?(10\.\d{4,}(?:\.\d+)*\/\S+)$", re.IGNORECASE)
+
+
+class DataSourceError(Exception):
+    """Exception raised for errors in data source handling."""
+
+    pass
+
+
+def _is_doi(source: str) -> bool:
+    """Check if source string is a DOI identifier.
+
+    Parameters
+    ----------
+    source : str
+        Source string to check.
+
+    Returns
+    -------
+    bool
+        True if source matches DOI pattern.
+    """
+    return bool(DOI_PATTERN.match(source))
+
+
+def _normalize_doi(source: str) -> str:
+    """Normalize DOI string by removing 'doi:' prefix if present.
+
+    Parameters
+    ----------
+    source : str
+        DOI string, possibly with 'doi:' prefix.
+
+    Returns
+    -------
+    str
+        Normalized DOI without prefix.
+    """
+    match = DOI_PATTERN.match(source)
+    if match:
+        return match.group(1)
+    return source
+
+
+def _resolve_doi(doi: str) -> str:
```

Comment:
Please use tom's code from today's call to get the zarr url: https://github.com/caltechlibrary/caltechdata_api/blob/main/caltechdata_api/download_file.py